creator-node/src/middlewares.js

```diff
@@ -56,47 +72,78 @@ async function syncLockMiddleware (req, res, next) {
   }
+  const userId = req.session.userId
 
-  let creatorNodeEndpoints
+  logPrefix = `${logPrefix} [userId = ${userId}]`
+
+  const selfSpID = nodeConfig.get('spID')
+  if (!selfSpID) {
+    return sendResponse(req, res, errorResponseServerError(`${logPrefix} Node failed to recover its own spID. Cannot validate user write.`))
+  }
+
+  /**
+   * Fetch current user replicaSetSpIDs
+   * Will throw error if selfSpID is not user primary
+   */
+  let replicaSetSpIDs
   try {
-    creatorNodeEndpoints = await getCreatorNodeEndpoints({
+    replicaSetSpIDs = await getReplicaSetSpIDs({
       serviceRegistry,
       logger: req.logger,
       wallet: req.session.wallet,
       blockNumber: req.body.blockNumber,
       ensurePrimary: true,
-      myCnodeEndpoint: serviceEndpoint
+      selfSpID,
+      userId
     })
   } catch (e) {
-    return sendResponse(req, res, errorResponseServerError(e))
+    return sendResponse(req, res, errorResponseServerError(`${logPrefix} ${e.message}`))
   }
-  const primary = creatorNodeEndpoints[0]
-
-  // Error if this node is not primary for user.
-  if (!primary || (primary && serviceEndpoint !== primary)) {
-    return sendResponse(
-      req,
-      res,
-      errorResponseUnauthorized(`This node (${serviceEndpoint}) is not primary for user. Primary is: ${primary}`)
-    )
+
+  const primarySpID = replicaSetSpIDs[0]
+
+  // TODO there should be no need to do this validation (standardize / cleanup errors between here and in getReplicaSetSpIDs())
+  if (selfSpID !== primarySpID) {
+    return sendResponse(req, res, errorResponseUnauthorized(`${logPrefix} Failed. This node's spID (${selfSpID}) does not match user's primary spID on chain (${primarySpID}).`))
   }
+
+  req.session.replicaSetSpIDs = replicaSetSpIDs
   req.session.nodeIsPrimary = true
-  req.session.creatorNodeEndpoints = creatorNodeEndpoints
 
-  req.logger.info(`ensurePrimaryMiddleware succeeded ${Date.now() - start} ms. creatorNodeEndpoints: ${creatorNodeEndpoints}`)
+  /**
+   * Convert replicaSetSpIDs to replicaSetEndpoints for later consumption (do not error on failure)
+   * Currently `req.session.creatorNodeEndpoints` is only used by `issueAndWaitForSecondarySyncRequests()`
+   * There is a possibility of failing to retrieve endpoints for each spID, so the consumer of req.session.creatorNodeEndpoints must perform null checks
+   */
+  const allRegisteredCNodes = await utils.getAllRegisteredCNodes(libs, logger)
+  const replicaSetEndpoints = replicaSetSpIDs.map(replicaSpID => {
+    if (replicaSpID === selfSpID) {
+      return nodeConfig.get('creatorNodeEndpoint')
+    }
+    const replicaSetInfo = (allRegisteredCNodes.filter(CNodeInfo => CNodeInfo.spID === replicaSpID))[0]
+    if (replicaSetInfo && replicaSetInfo.endpoint) {
+      return replicaSetInfo.endpoint
+    } else {
+      return null
+    }
+  })
+  req.session.creatorNodeEndpoints = replicaSetEndpoints
```

Comment:
does it make sense to have nulls in here? Should we call .filter(Boolean) after the map to remove them?

Reply:
they are removed later but you right, better to have that here. done